src/reason-parser/reason_parser.mly

```diff
@@ -882,34 +891,126 @@ type let_bindings =
   { lbs_bindings: Parsetree.value_binding list;
     lbs_rec: rec_flag;
     lbs_extension: (attributes * string Asttypes.loc) option;
+    lbs_combinator: Longident.t Asttypes.loc option;
     lbs_loc: Location.t }
 
-let mklbs ext rf lb loc =
+let mklbs combinator ext rf lb loc =
   { lbs_bindings = [lb];
     lbs_rec = rf;
     lbs_extension = ext;
+    lbs_combinator = combinator;
     lbs_loc = loc; }
 
 let addlbs lbs lbs' =
   { lbs with lbs_bindings = lbs.lbs_bindings @ lbs' }
 
 let val_of_let_bindings lbs =
   let str = Str.value lbs.lbs_rec lbs.lbs_bindings in
+  if lbs.lbs_combinator <> None then
+    raise Syntaxerr.(Error(Not_expecting(lbs.lbs_loc, "let.foo is not allowed at the top level")));
   match lbs.lbs_extension with
   | None -> str
   | Some ext -> struct_item_extension ext [str]
 
+(* Transforms
+
+     let.foo x = a
+     and y = b;
+     rest_of_code
+
+   into
+
+     Foo.let_(Foo.and_(a, b), (x, y) => rest_of_code);
+
+   foo is the "combinator." It is a module with two functions:
+
+   - Foo.let_ is a monadic bind or functor map operation, like Option.map,
+     Promise.then, etc. It applies a continuation (rest_of_code) to some value,
+     which is unwrapped from an option, promise, etc. It is used to replace let.
+   - Foo.and_ is a pairing operation, which takes two values, and wraps them in
+     an option, promise, etc. Foo.and_ is used to replace and. *)
+let combinator_call_of_let_bindings combinator let_bindings rest_of_code =
+  let combinator_loc = combinator.loc in
+  let combinator =
+    String.capitalize (String.concat "." (flatten combinator.txt)) in
```

Comment:
if the combinator is a longident (e.g. `let.Monads.opt x = y` this will cause issues.
I like the idea of allowing a longident if people wanted to namespace things, but given that we're doing auto-capitalization it might be needlessly confusing.
What do you thing?

Reply:
I think we will have to capitalize each component of the longident. I agree this is confusing, but we can basically say use a module name, and if you don't feel like capitalizing it, then don't.

Reply:
Actually, I agree. We can ban longidents from now, and allow only simple identifiers. We can restore support for longidents later on demand.